In [ ]:
!pip install gensim
!pip install python-Levenshtein

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 174.1/174.1 kB 5.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.0/3.0 MB 66.9 MB/s eta 0:00:00


<!DOCTYPE html>
<html>
<body>
  <h3>Gensim</h3>
  <p>Gensim is a Python library for topic modeling and document similarity analysis.</p>
  <h3>Levenshtein</h3>
  <p>Levenshtein distance is an algorithm used to measure the difference between two strings.</p>
</body>
</html>


***-------------------------------------------------------------------------------------------------------------------***

 The Amazon Cell Phones and Accessories Reviews dataset, is a valuable resource for sentiment analysis and customer feedback analysis. It contains reviews written by customers who have purchased cell phones and accessories on Amazon. Here's a brief introduction to this type of dataset

In [6]:
import gensim
import pandas as pd
df=pd.read_json("/content/drive/MyDrive/Cell_Phones_and_Accessories_5.json",lines=True)
df.head()

reviewerID        asin      reviewerName helpful  \
0  A30TL5EWN6DFXT  120401325X         christina  [0, 0]   
1   ASY55RVNIL0UD  120401325X          emily l.  [0, 0]   
2  A2TMXE2AFO7ONB  120401325X             Erica  [0, 0]   
3   AWJ0WZQYMYFQ4  120401325X                JM  [4, 4]   
4   ATX7CZYFXI1KW  120401325X  patrice m rogoza  [2, 3]   

                                          reviewText  overall  \
0  They look good and stick good! I just don't li...        4   
1  These stickers work like the review says they ...        5   
2  These are awesome and make my phone look so st...        5   
3  Item arrived in great time and was in perfect ...        4   
4  awesome! stays on, and looks great. can be use...        5   

                                     summary  unixReviewTime   reviewTime  
0                                 Looks Good      1400630400  05 21, 2014  
1                      Really great product.      1389657600  01 14, 2014  
2                             LOVE LOVE LOVE      1403740800  06 26, 2014  
3                                      Cute!      1382313600  10 21, 2013  
4  leopard home button sticker for iphone 4s      1359849600   02 3, 2013

In [7]:
df.shape

(194439, 9)

In [8]:
df.reviewText #we are interested in this column

0         They look good and stick good! I just don't li...
1         These stickers work like the review says they ...
2         These are awesome and make my phone look so st...
3         Item arrived in great time and was in perfect ...
4         awesome! stays on, and looks great. can be use...
                                ...                        
194434    Works great just like my original one. I reall...
194435    Great product. Great packaging. High quality a...
194436    This is a great cable, just as good as the mor...
194437    I really like it becasue it works well with my...
194438    product as described, I have wasted a lot of m...
Name: reviewText, Length: 194439, dtype: object

In [9]:
df.reviewText[5]

'These make using the home button easy. My daughter and I both like them.  I would purchase them again. Well worth the price.'

In [10]:
gensim.utils.simple_preprocess("These make using the home button easy. My daughter and I both like them.  I would purchase them again. Well worth the price.")

['these',
 'make',
 'using',
 'the',
 'home',
 'button',
 'easy',
 'my',
 'daughter',
 'and',
 'both',
 'like',
 'them',
 'would',
 'purchase',
 'them',
 'again',
 'well',
 'worth',
 'the',
 'price']

In [13]:
review_Text = df.reviewText.apply(gensim.utils.simple_preprocess)
#returns us a numpy array

*`import gensim
df.reviewText.apply(lambda x: gensim.utils.simple_preprocess(x))`*

We could have also used this function

In [17]:
print(review_Text)
print('----------------------------------------------------------')
print(review_Text[10])

0         [they, look, good, and, stick, good, just, don...
1         [these, stickers, work, like, the, review, say...
2         [these, are, awesome, and, make, my, phone, lo...
3         [item, arrived, in, great, time, and, was, in,...
4         [awesome, stays, on, and, looks, great, can, b...
                                ...                        
194434    [works, great, just, like, my, original, one, ...
194435    [great, product, great, packaging, high, quali...
194436    [this, is, great, cable, just, as, good, as, t...
194437    [really, like, it, becasue, it, works, well, w...
194438    [product, as, described, have, wasted, lot, of...
Name: reviewText, Length: 194439, dtype: object
----------------------------------------------------------
['this', 'case', 'fits', 'perfectly', 'on', 'the', 'and', 'keeps', 'me', 'powerd', 'all', 'day', 'can', 'complain', 'recommend', 'it', 'to', 'all']


In [19]:
model=gensim.models.Word2Vec(
    window=10,min_count=2,workers=4
)
model.build_vocab(review_Text,progress_per=1000)

In [21]:
print('Epochs : ',model.epochs)
print('Corpus Count : ',model.corpus_count)

Epochs :  5
Corpus Count :  194439


In [22]:
model.train(review_Text,total_examples=model.corpus_count,epochs=model.epochs)

(61504894, 83868975)

In [23]:
model.save("./word2vec-amazon-cell-accs-rev.model")

In [28]:
model.wv.most_similar("bad")

[('shabby', 0.6778111457824707),
 ('terrible', 0.6639342308044434),
 ('horrible', 0.596806526184082),
 ('good', 0.5800367593765259),
 ('awful', 0.5624942779541016),
 ('disappointing', 0.5253126621246338),
 ('crappy', 0.5233975648880005),
 ('ok', 0.517328679561615),
 ('funny', 0.5115740299224854),
 ('legit', 0.5104600191116333)]

In [29]:
model.wv.most_similar("good")

[('decent', 0.810697615146637),
 ('great', 0.7837907671928406),
 ('nice', 0.7169830799102783),
 ('fantastic', 0.7056553363800049),
 ('excellent', 0.6317567229270935),
 ('outstanding', 0.6120190620422363),
 ('superb', 0.6056107878684998),
 ('awesome', 0.6021170020103455),
 ('terrific', 0.6002836227416992),
 ('reasonable', 0.5855860114097595)]

The output received after executing `model.wv.most_similar("bad")` is a list of word-similarity pairs. Each pair consists of a word and a similarity score. The list represents the words that are most similar to the word "bad" based on the trained word vectors in the model.

Let's break down the output:
1. The word `"shabby"` has a similarity score of 0.6778111457824707, which means it is highly similar to the word "bad" in the context of the trained word vectors.

2. Similarly, `"terrible"` and `"horrible"` have high similarity scores of 0.6639342308044434 and 0.596806526184082, respectively.

3. On the other hand, `"good"` has a lower similarity score of 0.5800367593765259, indicating that it is less similar to `"bad"` compared to the previous words.

4. The words `"awful"`, `"disappointing"`, `"crappy"`, `"ok"`, `"funny"`, and `"legit"` also have similarity scores ranging from 0.5253126621246338 to 0.5104600191116333, indicating their similarity to the word "bad" but to a slightly lesser degree.

In [35]:
# Calculate and print similarity scores
similarity_score_1=model.wv.similarity(w1="cheap", w2="inexpensive")
similarity_score_2=model.wv.similarity(w1="good", w2="rubbish")
similarity_score_3=model.wv.similarity(w1="good", w2="product")
similarity_score_4=model.wv.similarity(w1="bad", w2="awful")
similarity_score_5=model.wv.similarity(w1="positive", w2="negative")
similarity_score_6=model.wv.similarity(w1="amazing", w2="dish")
print("Similarity between 'cheap' and 'inexpensive':", similarity_score_1)
print("Similarity between 'good' and 'rubbish':", similarity_score_2)
print("Similarity between 'good' and 'product':", similarity_score_3)
print("Similarity between 'bad' and 'awful':", similarity_score_4)
print("Similarity between 'positive' and 'negative':", similarity_score_5)
print("Similarity between 'amazing' and 'dish':", similarity_score_6)

Similarity between 'cheap' and 'inexpensive': 0.53239584
Similarity between 'good' and 'rubbish': 0.31537217
Similarity between 'good' and 'product': -0.032642324
Similarity between 'bad' and 'awful': 0.5624943
Similarity between 'positive' and 'negative': 0.72667813
Similarity between 'amazing' and 'dish': -0.05745638


1. A score close to 1 indicates a high similarity or relatedness between the words.
2. A score close to 0 suggests moderate similarity or unrelatedness.
3. A score close to -1 indicates a high dissimilarity or opposite meaning between the words.